### ISSUE TITLE
Grammar works as expected with Earley, parser fails with LALR

### ISSUE THREAD

**Issue by polwel** [erezsh/lark#22]
Below you find an adapted version of the calc example. I have explicitly added white space tokens, rather than ignoring white space in general.

It works with Earley, but not with LALR. As I understand, Lark should complain if the grammar is not parsable by LALR.

Note that I have defined a rule (token?) `_WS`. It corresponds to `WS_INLINE`, but the added underscore keeps it from appearing in the tree. `_WS` is capitalized, and should hence a terminal. On the other hand, it is not defined by either a string nor a regexp. Is it intentional that such a construct is allowed?

I tried Lark version 0.3.6.

```python
from lark import Lark, InlineTransformer

calc_grammar = """
    ?sum: product
        | sum [_WS] "+" [_WS] product   -> add
        | sum [_WS] "-" [_WS] product   -> sub

    ?product: atom
        | product [_WS] "*" [_WS]  atom  -> mul
        | product [_WS] "/" [_WS] atom  -> div

    ?atom: NUMBER           -> number
         | "-" atom         -> neg
         | "(" sum ")"

    %import common.NUMBER
    %import common.WS_INLINE

    _WS: WS_INLINE
"""


class CalculateTree(InlineTransformer):
    from operator import add, sub, mul, truediv as div, neg
    number = float


calc_parser = Lark(calc_grammar, parser='earley', start='sum')
# calc_parser = Lark(calc_grammar, parser='lalr', start='sum')


def calc(expr):
    tree = calc_parser.parse(expr)
    return CalculateTree().transform(tree)


if __name__ == '__main__':
    print(calc("1 + 3 * 4"))
```

**Comment by erezsh** [comment]
> I have explicitly added white space tokens, rather than ignoring white space in general.
Why? For what purpose?
LALR(1), as its name suggests, only has a look-ahead of 1 token. If you add whitespace to the grammar, you are effectively turning it into LR(0), which is significantly less powerful.

I should confess that there is a bug here, which I should fix. You can bypass it easily with:
    _WS.2: WS_INLINE
But as you'll see, it doesn't solve the bigger issue, of LALR's limited lookahead (which Earley, of course, doesn't have).

**Comment by polwel** [comment]
Hi, thanks for the quick answer.

I was suspecting that the grammar was no longer parsable by Earley. I just wanted to point out that Lark should have warned me about that. Your explanation makes sense to me.

> Why? For what purpose?

I was just playing around. Say I have a language where white spaces are generally ignored, except in a few constructs, e.g. in numbers with SI prefixes, like `12u`, standing for `12e-6`. Here whitespace between the numeral and the prefix are specifically disallowed.  Is there an easy way to do that other than polluting the grammar with optional white spaces, or capturing these numbers with a regexp, and then parsing them separately?

What exactly does `_WS.2: WS_INLINE` do? AFAICT, it doesn't help LALR, and it was working fine with Earley already.

**Comment by psboyce** [comment]
I have a language called SVF that I need to parse where LALR finds unexpected tokens even though the expected token would also work, and I'm also explicitly ignoring whitespace tokens. (I'm using LALR because SVF has C style comments and issue #24 precludes the use of the Earley parser) In SVF the last character in a statement _must_ be a semicolon, and the semicolon _must not_ be preceded by a whitespace character.

You explanation about why using explicit whitespace tokens causes trouble with LALR makes sense to me, but I'm not clear about how to work around this issue with LALR.

**Comment by erezsh** [comment]
Please see my comment on issue #24

**Comment by erezsh** [comment]
I'm closing this issue, since I fixed issue #24. If the problem persists feel free to re-open this issue, or open a new one.
